lib/babe/epoch.go

```diff
@@ -221,12 +222,34 @@ func (b *Service) incrementEpoch() (uint64, error) {
 // It returns the wrapped error errOverPrimarySlotThreshold
 // if it is not authorised.
 // output = return[0:32]; proof = return[32:96]
-func (b *Service) runLottery(slot, epoch uint64, epochData *epochData) (*VrfOutputAndProof, error) {
-	return claimPrimarySlot(
+func (b *Service) runLottery(slot, epoch uint64, epochData *epochData) (*types.PreRuntimeDigest, error) {
+	proof, err := claimPrimarySlot(
 		epochData.randomness,
 		slot,
 		epoch,
 		epochData.threshold,
 		b.keypair,
 	)
+	if err == nil {
+		preRuntimeDigest := types.NewBabePrimaryPreDigest(epochData.authorityIndex, slot, proof.output, proof.proof)
+
+		return preRuntimeDigest.ToPreRuntimeDigest()
+	}
+
+	if epochData.secondary == 0 {
+		return nil, err
+	}
+
+	if errors.Is(err, errOverPrimarySlotThreshold) {
+		proof, err = claimSecondarySlot(epochData.randomness, slot, epoch, epochData.authorities, b.keypair, epochData.authorityIndex)
+		if err != nil {
+			return nil, err
+		}
+		if proof != nil {
+			preRuntimeDigest := types.NewBabeSecondaryPlainPreDigest(epochData.authorityIndex, slot)
+			return preRuntimeDigest.ToPreRuntimeDigest()
+		}
+	}
+
+	return nil, err
```

Comment:
`claimSecondarySlot` doesn't return an error if we fail to claim a slot like `claimPrimarySlot` does, this causes the calling function `getFirstAuthoringSlot` to set the first authoring slot incorrectly as it will set it to a slot where we can't actually author as primary or secondary

Reply:
Yes,

When this comment was made, `claimSecondarySlot` did not return an error on failure. Thus, it would claim the slot. I have changed the code since then. And an error gets returned as expected.